compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -167,9 +164,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             "For closure={:?}, capture_information={:#?}",
             closure_def_id, delegate.capture_information
         );
+
         self.log_capture_analysis_first_pass(closure_def_id, &delegate.capture_information, span);
 
-        self.compute_min_captures(closure_def_id, capture_clause, delegate.capture_information);
+        let (capture_information, closure_kind, origin) = self
```

Comment:
I think this code would be cleaner if we made an `upvar` module with submodules like:

* `rustc_typeck::check::upvar::collect_uses`
* `rustc_typeck::check::upvar::minimization`

and just had the "top-level logic" in `rustc_typeck::check::upvar`.

I also think we should include a comment with the "abstract algorithm" -- but maybe it should live in the reference, actually, and we should just reference to it from here.

Reply:
(Let's break this out into a separate cleanup issue)

Reply:
Done: https://github.com/rust-lang/project-rfc-2229/issues/55

Reply:
Yea I was thinking of referencing the Rust Reference here than writing the algoirthm out

Reply:
https://github.com/rust-lang/rust/blob/efeea79a6bb6867b220471e0396b82b6b68ebc4c/compiler/rustc_typeck/src/check/upvar.rs#L1495-L1503

Reply:
It's this peice of code